apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/ForwardIncludeRequestDispatcherInstrumentation.java

```diff
@@ -0,0 +1,111 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 - 2019 Elastic and contributors
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.servlet;
+
+import co.elastic.apm.agent.bci.ElasticApmInstrumentation;
+import co.elastic.apm.agent.impl.transaction.Span;
+import co.elastic.apm.agent.impl.transaction.TraceContextHolder;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.NamedElement;
+
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+import javax.annotation.Nullable;
+import javax.servlet.RequestDispatcher;
+import javax.servlet.ServletRequest;
+import javax.servlet.http.HttpServletRequest;
+import java.util.Arrays;
+import java.util.Collection;
+
+import static net.bytebuddy.matcher.ElementMatchers.*;
+
+public class ForwardIncludeRequestDispatcherInstrumentation extends ElasticApmInstrumentation {
+
+    private static final String SPAN_TYPE_REQUEST_DISPATCHER_FORWARD = "servlet.request-dispatcher.forward";
+    private static final String FORWARD = "FORWARD";
+    private static final String SPAN_TYPE_REQUEST_DISPATCHER_INCLUDE = "servlet.request-dispatcher.include";
+    private static final String INCLUDE = "INCLUDE";
+
+    @Override
+    public Class<?> getAdviceClass() {
+        return ForwardIncludeRequestDispatcherAdvice.class;
+    }
+
+    public static class ForwardIncludeRequestDispatcherAdvice extends ForwardIncludeRequestDispatcherInstrumentation {
+        @Advice.OnMethodEnter(suppress = Throwable.class)
+        private static void beforeExecute(@Advice.Local("span") @Nullable Span span,
+                                          @Advice.Origin("#m") String method,
+                                          @Advice.Argument(0) @Nullable ServletRequest request) {
+            if (tracer == null || tracer.getActive() == null) {
+                return;
+            }
+            final TraceContextHolder<?> parent = tracer.getActive();
```

Comment:
The instrumented RequestDispatcher might call another, delegate RequestDispatcher which would lead to the creation of two identical spans. See for example https://github.com/spring-projects/spring-security/blob/9c1eac79e2abb50f7b01e77c2418566f2a30532f/web/src/main/java/org/springframework/security/web/firewall/RequestWrapper.java#L151-L155.

It's a tough one to detect, because there can be legitimate cases where you have a `FORWARD` span as a direct parent of another `FORWARD` span. 🤔

Reply:
You could probably check if the parent span has the same name as this span would have.

Reply:
@felixbarny how access to parentSpan via  
```
final TraceContextHolder<?> parent = tracer.getActive();?
```

Reply:
You can do an `instanceof AbstractSpan` check and cast